### DISCUSSION TITLE
What is the correct way to use `var body: some ReducerProtocol` in Swift 5.6?

### DISCUSSION THREAD

**Discussion by ylorn** [pointfreeco/swift-composable-architecture#1665] [Q&A] [answered]
Following code does not compile with Xcode 13.4 which embeds Swift 5.6.1:

```swift
struct Demo: ReducerProtocol {

    struct State {
        var value: String
    }

    enum Action {
        case task
    }

    @ReducerBuilder<State, Action>
    var body: some ReducerProtocol {

        Reduce { state, action in
            switch action {
                case .task:
                    return .none
            }
        }
    }
}
```

It requires `func reduce(into state: inout State, action: Action) -> EffectTask<Action>` to be implemented even if `var body` property presents.

**Comment by mbrandonw** [comment, marked as answer]
Hi @ylorn, since Swift 5.6.1 doesn't support primary associated types you must structure the `body` like this:

```swift
struct Demo: ReducerProtocol {
  struct State {
    var value: String
  }
  enum Action {
    case task
  }

  var body: Reduce<State, Action> {
    Reduce { state, action in
      switch action {
      case .task:
        return .none
      }
    }
  }
}
```

We have more information in our `ReducerProtocol` [migration guide](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/migratingtothereducerprotocol), in particular the [section on Swift 5.6](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/migratingtothereducerprotocol#Migration-using-Swift-56).

**Reply by mbrandonw** [reply]
Quick note that the way this works is that secretly under the hood the result builders will do type erasure, and this is why if you try to compile this style of code in Swift 5.7+ you will get a deprecation warning letting you know that you should upgrade to the `some ReducerProtocol<State, Action>` style.

**Reply by ylorn** [reply]
I wasn't aware of the migration notes, Thanks!
